### ISSUE TITLE
confuse about the attn_weight of the _call_dense method of GATConv layer

### ISSUE THREAD

**Issue by LJingnan** [danielegrattarola/spektral#205]
I notice that `mask = -10e9 * (1.0 - a)` is used to downgrade those with limited coef, but, as you have set
`        a = tf.linalg.set_diag(a, tf.ones(shape, a.dtype))`, the coef-matrix's diagnol is set to 1, which means 
the mask of the diagnol(self attn) is 0. Think about a pair coef is 0.9, which is high enough, its mask will be
-10e-8. When use softmax latter, the 0.9 coef will be set to 0.

`    def _call_dense(self, x, a):
        shape = tf.shape(a)[:-1]
        a = tf.linalg.set_diag(a, tf.zeros(shape, a.dtype))
        a = tf.linalg.set_diag(a, tf.ones(shape, a.dtype))
        x = tf.einsum("...NI , IHO -> ...NHO", x, self.kernel)
        attn_for_self = tf.einsum("...NHI , IHO -> ...NHO", x, self.attn_kernel_self)
        attn_for_neighs = tf.einsum(
            "...NHI , IHO -> ...NHO", x, self.attn_kernel_neighs
        )
        attn_for_neighs = tf.einsum("...ABC -> ...CBA", attn_for_neighs)

        attn_coef = attn_for_self + attn_for_neighs
        attn_coef = tf.nn.leaky_relu(attn_coef, alpha=0.2)

        mask = -10e9 * (1.0 - a)
        attn_coef += mask[..., None, :]
        attn_coef = tf.nn.softmax(attn_coef, axis=-1)
        attn_coef_drop = self.dropout(attn_coef)

        output = tf.einsum("...NHM , ...MHI -> ...NHI", attn_coef_drop, x)

        return output, attn_coef`

**Comment by danielegrattarola** [comment]
I am not sure I see what you mean. 

`mask = -10e9 * (1.0 - a)` changes `a` so that all the `1`s become `0`s and all the `0`s become `-10e9`.

Then you add the mask to `attn_coef` so that all attention coefficients that are associated with a `1` in the adjacency matrix (including the diagonal) remain unchanged (since they are summed with 0) and all the attention coefficients associated with a 0 become `-10e9 + attn_coef[i, j]` which will be close to `-10e9`.

Finally, the softmax will behave as if the values of `attn_coef` that were masked to `-10e9` do not exist. 

Am I missing something?

**Comment by LJingnan** [comment]
Yes, but if there is a coef array like [0, 0.9, 1] for example. After mask, it would become [-10e9, -10e8 + 0.9, 0 + 1]，
after, softmax, the weight will become [0,0,1], which is strange because coef is as high as 0.9.

**Comment by danielegrattarola** [comment]
In that case, it would mean that the associated row of the adjacency matrix is [0, 0, 1], so it is expected that the second attention coefficient is 0.
Most of the `attn_coef` entries will be non-zero, since `attn_coef` is the result of a dense multiplication.

**Comment by LJingnan** [comment]
I think I get your point. Do you mean that the element of adjacency matrix should be either 1 or 0.

**Comment by danielegrattarola** [comment]
Ah sorry, I must have misunderstood that point!
Yes, GAT expects a binary adjacency because of this reason.

**Comment by LJingnan** [comment]
Thank you so much, I read the paper again, and you are right. As I have some pre-define coef-matrix (some industry knowledge), I think I can make a little change to adjust for it, like removing the mask. May I ask your suggestion for how to combine self-attention and industrial knowledge? Again, thanks for your excellent work!

**Comment by danielegrattarola** [comment]
If you remove the mask altogether, then all the zeros in the adjacency matrix will be considered as edges.

If you are willing to code your own layer you can try to binarize the adjacency matrix only for computing the mask, so something like this: 

```
mask = -10e9 * (1.0 - tf.cast(a > 0, tf.float32))
attn_coef += mask[..., None, :]
attn_coef = tf.nn.softmax(attn_coef, axis=-1)
attn_coef *= a
```

but that is a bit more expensive.

**Comment by LJingnan** [comment]
I get it. Thank you very very much.
